[CIRCLE-20816] Document matrix jobs (#4247)
circleci/circleci-docs: jekyll/_cci2/configuration-reference.md

For more information, see the [Executing Workflows For a Git Tag]({{ site.baseurl }}/2.0/workflows/#executing-workflows-for-a-git-tag) section of the Workflows document.

###### **`matrix`**
A matrix job allows you to run a parameterized job multiple times with different
arguments.

Key | Required | Type | Description
----|----------|------|------------
parameters | Y | Map  | A map of parameter names to every value the job should be called with
exclude | N | Map | A list of argument maps that should be excluded from the matrix
alias | N | String | An alias for the matrix, usable from another job's `requires` stanza. Defaults to the name of the job being executed

The following is a basic example of using matrix jobs. All features are
explained in more detail below.

```yaml
workflows:
  workflow:
    jobs:
      - build:
        matrix:
          parameters:
            version: ["0.1", "0.2", "0.3"]
            platform: ["macos", "windows", "linux"]

> ```

Reviewer: This is defo helpful 👍

###### Excluding sets of parameters from a matrix
{:.no_toc}
Sometimes you may wish to run a job with every combination of arguments _except_
some value or values. You can use an `exclude` stanza to achieve this:

```yaml
workflows:
  workflow:
    jobs:
      - build:
          matrix:
            parameters:
              a: [1, 2, 3]
              b: [4, 5, 6]
            exclude:
              - a: 3
                b: 5
```

The matrix above would result in 8 jobs being run: every combination of the
parameters `a` and `b`, excluding `{a: 3, b: 5}`

###### Requiring an entire matrix
{:.no_toc}
Another job can `require` an entire matrix (that is, every job in the matrix)
using its `alias`. The `alias` defaults to the name of the job being invoked.
```yaml
workflows:
  workflow:
    jobs:
      - deploy:
          matrix:
            parameters:
              version: ["0.1", "0.2"]
      - another-job:
          requires:
            - deploy
```

This means that `another-job` will require both `deploy-0.1` and `deploy-0.2` to
finish before it runs. 

> You can also manually set the `alias` to a custom value. For example:

Reviewer 2: Do you think this would help explain when the alias is useful, or just complicate things because the following example doesn't demonstrate any ambiguity being addressed?
```suggestion
You can also manually set the `alias` to a custom value, to clear up any ambiguity between job names.
For example:
```

```yaml
workflows:
  workflow:
    jobs:
      - deploy:
          matrix:
            parameters:
              version: ["0.1", "0.2"]
            # setting a custom alias
            alias: my-deploy-matrix
      - another-job:
          requires:
            # using a custom alias
            - my-deploy-matrix
```

###### Matrix parameters
{:.no_toc}
When running a job in a matrix, other job arguments (such as `requires` or
`name`) can contain matrix parameters, in the form `<< matrix.[parameter-name] >>`
For example:
```yaml
workflows:
  workflow:
    jobs:
      - build-0.1: ...
      - build-0.2: ...
      - deploy:
          matrix:
            parameters:
              version: ["0.1", "0.2"]
          name: "deploy-<< matrix.version >>"
          requires:
            - "build-<< matrix.version >>"
```

This matrix results in two jobs being run. One, called `deploy-0.1`,
would require the `build-0.1` job. The other, called `deploy-0.2`, would require
the `build-0.2` job.

In a typical use case, you may wish to require a single job from a matrix from
within another matrix. To do so, use matrix parameters to set the `name` of the

> required job, and the `requires` stanza of the dependent job:

Reviewer 2: Idea: "requiring an entire matrix" and "requiring matrix jobs from within a matrix" feel like similar enough topics that they could be grouped together into a section called `Dependencies and matrix jobs` that showcases the two use cases people might be interested in.

```yaml
workflows:
  workflow:
    jobs:
      - build:
          # provide a name to each job in this matrix
          name: "build-<< matrix.os >>-<< matrix.version >>"
          matrix:
            parameters:
              version: ["0.1", "0.2"]
              os: ["windows", "linux"]
      - deploy:
          matrix:
            parameters:
              version: ["0.1", "0.2"]
              os: ["windows", "linux"]
          requires:
            # require the above name in each job in this matrix

>             - "build-<< matrix.os >>-<< matrix.version >>"

Reviewer: Does this mean, if `deploy` requires a single build job, you would use the specific parameters in the requires naming? Like: `requires: -"build-windows-0.2"` for example? Or is this just showing the full `deploy` matrix relies on the full `build` matrix?

Author: There are three possible scenarios:
- _each_ deploy job depends on exactly one `build` job. So the `deploy-windows-0.2` job depends on `build-windows-0.2`, `deploy-macos-0.1` depends on `build-macos-0.1`, etc. That's this example ^

- all `deploy` jobs depends on a single `build` job. In that case the `requires` wouldn't have any templating at all, it's just be `requires: [my-job-name]`

- we don't want to run any deploy jobs unless ALL the build jobs succeeded. Then we will require the whole build matrix, maybe using an alias: `requires: [my-build-matrix-alias]`

Does that answer your question?

Reviewer: Yes I think so thank you!

exclude | N | Map | A list of argument maps that should be excluded from the matrix
alias | N | String | An alias for the matrix, usable from another job's `requires` stanza. Defaults to the name of the job being executed

The following is a basic example of using matrix jobs. All features are
explained in more detail below.

```yaml
workflows:
  workflow:
    jobs:
      - build:
          # In the invocation of a matrix job, you may use the special
          # template variables `<< matrix.* >>`. These will expand to
          # the value of the parameter for _each_ job within the matrix.

          # Naming each job in the matrix can be useful if you want to
          # require a particular matrix job from another job. The name
          # defaults to the job name ("build" here) concatenated with
          # all of the matrix values, sorted by their names in
          # alphabetical order.

          # The default name for this example would be:
          # name: "build-<< matrix.platform >>-<< matrix.version >>"

          name: "build-v<< matrix.version >>-for-<< matrix.platform >>"

          # You may also require jobs. You can even use this to require a

>           # particular matrix job from another matrix job!

Reviewer 2: Might be clearer to refer to the "cell" of a matrix job as just a "job"?
```suggestion
          # particular job in another matrix job!
```

          requires:
            - lint
            - "prepare-v<< matrix.version >>-for-<< matrix.platform >>"

          matrix:
            # The job will be invoked once for each combination of parameters.
            # In this case, the job would be invoked 9 times, once for each
            # version/platform combination.
            parameters:
              version: ["0.1", "0.2", "1.0"]
              platform: ["windows", "linux", "macos"]

            # You may provide a matrix alias in order to require the entire
            # matrix later. The default alias is the name of the job being
            # invoked, which in this case would be:
            # alias: build
            alias: my-matrix

            # You may `exclude` some sets of parameters from the matrix.
            exclude:
              - version: 0.2
                platform: linux
              - version: 1.0
                platform: macos

> ```

Reviewer 2: To me, this is the "basic example" of matrices:

```yaml
- build:
    matrix:
      parameters:
        version: ["0.1", "0.2", "1.0"]
        platform: ["windows", "linux", "macos"]
```

If you really want a full comprehensive example as well, for people to refresh their memory on the syntax, it would be more on-brand with the rest of our documentation if that came at the end of the matrix section and didn't have as many comments.

Author: This makes sense - I removed the "comprehensive example."

###### Matrix parameters
{:.no_toc}
When running a job in a matrix, other job arguments (such as `requires` or
`name`) can contain matrix parameters, in the form `<< matrix.[parameter-name] >>`
For example:

> ```

Reviewer: Could you set each code block to yaml?

Author: Yep! Done

> - deploy:

Reviewer: Could we add in the workflows key, or jobs key here to show where in the config this is? It's not immediately obvious to me.

Reviewer: Same for all the examples below... 😄

Author: Yep, sounds good! :+1:

    matrix:
      parameters:
        version: ["0.1", "0.2"]
    name: "deploy-<< matrix.version >>"
    requires:
      - "build-<< matrix.version >>"
```

> This matrix result in two jobs being run. One, called `deploy-0.1`,

Reviewer: ```suggestion
This matrix results in two jobs being run. One, called `deploy-0.1`,
```

      exclude:
        - a: 3
          b: 5
```

The matrix above would result in 8 jobs being run: every combination of the
parameters `a` and `b`, excluding `{a: 3, b: 5}`

###### Requiring an entire matrix
{:.no_toc}
Another job can `require` an entire matrix (that is, every job in the matrix)
using its `alias`. The `alias` defaults to the name of the job being invoked.
```
- deploy:
    matrix:
      parameters:
        version: ["0.1", "0.2"]
- another-job:
    requires:
      - deploy
```

This means that `another-job` will require both `deploy-0.1` and `deploy-0.2` to
finish before it runs. 

You can also manually set the `alias` to a custom value. For example:

```
- deploy:
    matrix:
      parameters:
        version: ["0.1", "0.2"]
      # setting a custom alias
      alias: my-deploy-matrix
- another-job:
    requires:
      # using a custom alias
      - my-deploy-matrix
```

###### Requiring a single job from a matrix
{:.no_toc}
Every job in a matrix has a name. By default, the name will be the job's name,
concatenated with the value of each matrix parameter alphabetized by the
*parameter name*, separated by hyphens. These names can be used to `require` a

> job, including from within another matrix:

Reviewer 2: Going into depth on how the default name is generated feels a bit TMI to me, since it's meant to be specified by the user if they have a specific opinion on what it should look like.

Author: I was thinking that it'd be more comprehensible if I could explain how to require a job _without_ specifically naming it... but that seems silly. I removed this part and just explicitly named the job and required that.

```
- build:
    matrix:
      parameters:
        version: ["0.1", "0.2"]
        os: ["windows", "linux"]
- deploy:
    matrix:
      parameters:
        version: ["0.1", "0.2"]
        os: ["windows", "linux"]
    requires:
      # using matrix parameters to require the matching `build` job
      - "build-<< matrix.os >>-<< matrix.version >>"
```

> The `name` of each job within the matrix can also be set manually. It's best to

Reviewer: ```suggestion
The `name` of each job within the matrix can also be set manually. It is best to
```
